File: app/src/main/java/ru/e6atb/chat/ForwardMessageFormatter.java
package ru.e6atb.chat;

final class ForwardMessageFormatter {
	static final int MAX_MESSAGE_BYTES = 4096;

	private ForwardMessageFormatter() {
	}

	static String compose(String header, String body) {
		String safeHeader = header == null ? "" : header.trim();
		String safeBody = body == null ? "" : body.trim();
		if (safeHeader.length() == 0) return safeBody;
		if (safeBody.length() == 0) return safeHeader;
		return safeHeader + "\n\n" + safeBody;
	}

	static boolean fitsServerLimit(String value) {
		return utf8Length(value) <= MAX_MESSAGE_BYTES;
	}

	static int utf8Length(String value) {
		if (value == null || value.length() == 0) return 0;
		int bytes = 0;
		for (int i = 0; i < value.length(); i++) {
			char ch = value.charAt(i);
			if (ch <= 0x7f) {
				bytes += 1;
			} else if (ch <= 0x7ff) {
				bytes += 2;
			} else if (Character.isHighSurrogate(ch)
					&& i + 1 < value.length()
					&& Character.isLowSurrogate(value.charAt(i + 1))) {
				bytes += 4;
				i++;
			} else {
				bytes += 3;
			}
		}
		return bytes;
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/ForwardMessageFormatter.kt
package ru.e6atb.chat

object ForwardMessageFormatter {
    const val MAX_MESSAGE_BYTES = 4096

    @JvmStatic fun compose(header: String?, body: String?): String {
        val safeHeader = header.orEmpty().trim()
        val safeBody = body.orEmpty().trim()
        return when {
            safeHeader.isEmpty() -> safeBody
            safeBody.isEmpty() -> safeHeader
            else -> "$safeHeader\n\n$safeBody"
        }
    }

    @JvmStatic fun fitsServerLimit(value: String?) = utf8Length(value) <= MAX_MESSAGE_BYTES

    @JvmStatic fun utf8Length(value: String?): Int {
        if (value.isNullOrEmpty()) return 0
        var bytes = 0
        var index = 0
        while (index < value.length) {
            val character = value[index]
            bytes += when {
                character.code <= 0x7f -> 1
                character.code <= 0x7ff -> 2
                Character.isHighSurrogate(character) && index + 1 < value.length && Character.isLowSurrogate(value[index + 1]) -> { index++; 4 }
                else -> 3
            }
            index++
        }
        return bytes
    }
}
